x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -97,12 +107,35 @@ final void startComputeOnDataNodes(
                         nodePermits.putIfAbsent(node, new Semaphore(1));
                     }
                 }
-                pendingShardIds.addAll(targetShards.shards.keySet());
+                pendingShardIds.addAll(order(targetShards));
                 trySendingRequestsForPendingShards(targetShards, computeListener);
             }
         }, listener::onFailure));
     }
 
+    private static List<ShardId> order(TargetShards targetShards) {
+        var ordered = new ArrayList<>(targetShards.shards.keySet());
+        ordered.sort(Comparator.comparingInt(shardId -> nodeOrder(targetShards.getShard(shardId).remainingNodes)));
+        return ordered;
+    }
+
+    private static int nodeOrder(List<DiscoveryNode> nodes) {
+        if (nodes.isEmpty()) {
+            return Integer.MAX_VALUE;
+        }
+        // assumes all shard nodes have same roles
+        var node = nodes.getFirst();
+        if (node.hasRole(DiscoveryNodeRole.DATA_CONTENT_NODE_ROLE.roleName()) || node.hasRole(DiscoveryNodeRole.DATA_ROLE.roleName())) {
+            return 0;
+        }
```

Comment:
I wonder if these should just be in the list. Or maybe if the list should be tuples.

Reply:
It'd be useful to put a comment wherever we handle them and call them "effectively hot tier" with a little description of the stuff at https://www.elastic.co/guide/en/elasticsearch/reference/current/node-roles-overview.html#data-content-node . I'm not used to the CONTENT role at least.